ext/opentelemetry-ext-postgresql/src/opentelemetry/ext/postgresql/__init__.py

```diff
@@ -0,0 +1,94 @@
+# Copyright 2019, OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+The opentelemetry-ext-postgresql package allows tracing PostgreSQL queries made by the
+Psycopg2 library.
+"""
+
+import typing
+
+import psycopg2
+from psycopg2.extensions import cursor as pgcursor
+import wrapt
+
+from opentelemetry.context import Context
+from opentelemetry.trace import SpanKind, Tracer
+from opentelemetry.trace.status import Status, StatusCanonicalCode
+
+QUERY_WRAP_METHODS = ["execute", "executemany", "callproc"]
+DATABASE_COMPONENT = "postgresql"
+DATABASE_TYPE = "sql"
+
+
+def trace_integration(tracer):
+    """Integrate with PostgreSQL Psycopg library.
+       Psycopg: http://initd.org/psycopg/
+    """
+
+    if tracer is None:
+        raise ValueError("The tracer is not provided.")
+
+    # pylint: disable=unused-argument
+    def wrap(wrapped, instance, args, kwargs):
+        """Patch Psycopg connect method.
+        """
+        connection = wrapped(*args, **kwargs)
+        connection.cursor_factory = TraceCursor
+        return connection
+
+    wrapt.wrap_function_wrapper(psycopg2, "connect", wrap)
+
+    class TraceCursor(pgcursor):
```

Comment:
Why doesn't this use opentelemetry-ext-dbapi from #264? Did you run into the issue I described in https://github.com/open-telemetry/opentelemetry-python/pull/264/files#r348452864? In any case, instead of deriving from pgcursor, you might want to use a wrapt ObjectProxy https://wrapt.readthedocs.io/en/latest/wrappers.html#object-proxy. That way, you could implement a generic solution. Or you could add support for such cursor_factory instrumentations to ext-dbapi.

Another question: What happens if the user manually specifies/sets cursor_factory? Will either our or the user's cursor_factory be ignored?

Reply:
I based this in current OpenCensus implementation and started working with this before adding dbapi in other PR, makes sense to reuse it as much as possible, cursor_factory in fact will be overriden like you mentioned and is current behavior in OpenCensus, will file a bug for it and ensure it doesn't happen here

Reply:
cursor_factory is a not thing in the Python DB API specification, so I guess it doesn't makes sense to add it in dbapi ext, dbapi ext is currently wrapping connect method, the actual connection and the cursor, in psycopg all of these are implemented in C, that is why we are hooking up to the cursor_factory, so I'm still not sure how we can reuse dbapi ext here, any suggestion would be appreciated.

Using ObjectProxy instead of pgcursor sounds like a good approach

Reply:
I think this could be an extension of the DBAPI PR. looking at the chunk that actually starts a span and sets values, it's very similar to the DBApi functionality.

Although one caveat as a con is the fact that the API is slightly different in that psycopg2 accepts sql.SQL objects rather than just strings, but that could reconciled with a light wrapper on all of them.

ultimately this is an auto-instrumentation though, so as long as we don't expose APIs that people may rely on, we can probably feel safe marking this as a potential refactor.

Reply:
Updated the code to use dbapi ext as much as possible, including span attributes related to connection population and actual span creation functionality